astronomy/standard_product_3.cpp

```diff
@@ -309,30 +366,76 @@ StandardProduct3::StandardProduct3(
         }
       }
 
-      orbit.Append(epoch, {position, velocity});
+      // Bad or absent positional and velocity values are to be set to 0.000000.
+      if (position == ITRS::origin || velocity == Velocity<ITRS>()) {
+        if (!arc.Empty()) {
+          orbit.push_back(make_not_null_unique<DiscreteTrajectory<ITRS>>());
+        }
+      } else {
+        arc.Append(epoch, {position, velocity});
+      }
     }
   }
   if (dialect != Dialect::ILRSA) {
     CHECK_EQ(columns(1, 3), "EOF") << location;
     read_line();
   }
+  for (auto& [id, orbit] : orbits_) {
+    // Do not leave a final empty trajectory if the orbit ends with missing
+    // data.
+    if (orbit.back()->Empty()) {
+      orbit.pop_back();
+    }
+  }
   CHECK(!line.has_value()) << location;
+  if (!has_velocities_) {
+    for (auto& [id, orbit] : orbits_) {
+      for (auto& arc : orbit) {
+        switch (arc->Size()) {
+#define COMPUTE_VELOCITIES_CASE(n)    \
+          case n:                             \
+            arc = ComputeVelocities<n>(*arc); \
+            break
+          COMPUTE_VELOCITIES_CASE(1);
+          COMPUTE_VELOCITIES_CASE(2);
+          COMPUTE_VELOCITIES_CASE(3);
+          COMPUTE_VELOCITIES_CASE(4);
+          COMPUTE_VELOCITIES_CASE(5);
+          COMPUTE_VELOCITIES_CASE(6);
+          COMPUTE_VELOCITIES_CASE(7);
+          COMPUTE_VELOCITIES_CASE(8);
+#undef COMPUTE_VELOCITIES_CASE
+          default:
```

Comment:
This is weird.  Why not a normal `9` branch and a `LOG(FATAL)` here?

Reply:
There can be (and most often will be) more than 9 points in the orbit.
The template parameter, being the number of points used by the finite difference formula, must be smaller than or equal to the number of points in the orbit, but cannot be arbitrarily large (we only have coefficients for up to 9-point difference formulæ).